### ISSUE TITLE
cors: true

### ISSUE THREAD

**Issue by nalbion** [awslabs/serverless-application-model#23]
One thing that I like about serverless.com is that CORS can be enabled for an endpoint in one short line:

    cors: true

Currently, to enable CORS using `x-amazon-apigateway-integration`, it is necessary to:

- add a header to each response:

```yaml
responses:
        '200':
          description: Request successful
          headers:
            Access-Control-Allow-Origin:
              type: string
          schema:
            $ref: '#/definitions/Success'
        '400':
          description: Authentication failed
          headers:
            Access-Control-Allow-Origin:
              type: string
          schema:
            $ref: '#/definitions/Error'
```

- add responseParameters in `x-amazon-apigateway-integration`:

```yaml
default:
            statusCode: "200"
            responseParameters:
              method.response.header.Access-Control-Allow-Origin : "'*'"
          '.*"status":400.*':
            statusCode: '400'
            responseParameters:
              method.response.header.Access-Control-Allow-Origin : "'*'"
```

- add a massive `options` section:
```yaml
options:
      summary: CORS support
      description: |
        Enable CORS by returning correct headers
      responses:
        200:
          description: Default response for CORS method
          headers:
            Access-Control-Allow-Headers:
              type: string
            Access-Control-Allow-Methods:
              type: string
            Access-Control-Allow-Origin:
              type: string
      x-amazon-apigateway-integration:
        type: mock
        requestTemplates:
          application/json: |
            {"statusCode" : 200}
        responses:
          "default":
            statusCode: "200"
            responseParameters:
              method.response.header.Access-Control-Allow-Headers : "'Content-Type,X-Amz-Date,Authorization,X-Api-Key'"
              method.response.header.Access-Control-Allow-Methods : "'*'"
              method.response.header.Access-Control-Allow-Origin : "'*'"
            responseTemplates:
              application/json: |
                {}
```

...it'd be great if `Transform: AWS::Serverless-2016-10-31` could transform the swagger schema, automatically adding all of this if `cors: true` is present in the CloudFormation object

**Comment by sanathkr** [comment]
I like this. This would be doable for implicit APIs specified in the Events section of a Function resource, because CloudFormation generates the swagger for you. For APIs created using AWS::Serverless::API resource, users will supply their own Swagger file in S3.  The logic that parses Serverless resources does not have the access to read and manipulate S3 files.

**Comment by nalbion** [comment]
Does the former allow you to save a local copy of the swagger schema, so that we may merge/cross-check the generated and documenting swagger schema?

**Comment by sanathkr** [comment]
Yeah. After you create the stack you can download the generated template using CloudFormation [get-template](http://docs.aws.amazon.com/cli/latest/reference/cloudformation/get-template.html) command. It will have the Swagger template embedded in it

**Comment by dinvlad** [comment]
This could be somewhat DRYed by defining the default response at the root of the Swagger file:
```
x-amazon-apigateway-integration-cors-headers:
  method.response.header.Access-Control-Allow-Methods: "'DELETE,GET,HEAD,OPTIONS,PATCH,POST,PUT'"
  method.response.header.Access-Control-Allow-Headers: "'Content-Type,Authorization,X-Amz-Date,X-Api-Key,X-Amz-Security-Token'"
  method.response.header.Access-Control-Allow-Origin: "'*'"
```
After that, these headers can be referenced in `x-amazon-apigateway-integration:` through
```
responseParameters:
  allOf:
    - $ref: '#/x-amazon-apigateway-integration-cors-headers'
```

**Comment by nalbion** [comment]
Personally, I'm a fan of Swagger, but `x-amazon-apigateway-integration` with CORS adds too much noise.

**Comment by dinvlad** [comment]
It'd be nice to keep an option to specify a Swagger template, but without all the extra clutter. One case we need to use Swagger for is for input validation, where we'd like to make sure the shape of the API is preserved between that defined in a SAM template and that specified explicitly in Swagger (with the latter having extra information about input parameter types etc.)

Could there be an option to define a "template template" Swagger, with parameters specified over a `${SOME_VAR}` syntax that are then dynamically expanded during processing of the SAM template? This will be then very similar to how currently the SAM template is itself "expanded" into a normal CloudFormation template during packaging.

**Comment by frehner** [comment]
To echo what @rabowskyb said, it would be nice to have a `cors: true` option in the `sam.yml` file itself. I'm using SAM to build a lot of small microservices and a swagger file seems overkill for something that literally only has one endpoint.

**Comment by sanathkr** [comment]
Here is a proposal of how this could work. API Event will get a new property called `Cors`. Here are two ways to use it:

## Shorthand
```
Type: AWS::Serverless::Function
Properties:
  Runtime: ...
  Events:
     GetApi:
         Type: Api
         Properties:
              Cors: "*"
             Method: GET
             Path: /foo

     PostApi:
         Type: Api
         Properties:
              Cors: "http://example.com"
             Method: POST
             Path: /foo
```

This will generate CORS headers automatically, for each Path+Method combo and set Allow-Origin to "\*".  The value of Cors property is a list of origins you want to allow. Allow-Headers is always set to "\*".

## Long Form
```
Type: AWS::Serverless::Function
Properties:
  Runtime: ...
  Events:
     GetApi:
         Type: Api
         Properties:
              Cors: 
                  AllowOrigin: "List of origins"
                  AllowHeaders: "List of allowed headers"
             Method: GET
             Path: /foo

     PostApi:
         Type: Api
         Properties:
              Cors: 
                  AllowOrigin: "List of origins"
                  AllowHeaders: "List of allowed headers"
             Method: POST
             Path: /foo
```

This will generate CORS headers automatically, for each Path+Method combo and set Allow-Origin and Allow-Headers to whatever value that was specified. 

Would this address use-cases you guys are going after? 

Unfortunately we can't add the same functionality to `AWS::Serverless::Api` property because the translator converting SAM to CloudFormation cannot read and modify your Swagger file.

**Comment by rabowskyb** [comment]
So both the short form and long form will be offered?  Would it be possible to allow use of wildcards in the list of origins, for example "*.example.com"?  

Overall, your proposal above appears to eliminate the need to use swagger to enable CORS, while providing a degree of flexibility to specify allowed origins and headers.  This supports my use case of webapps backed by serverless backends deployed with SAM.

**Comment by sanathkr** [comment]
yes, both forms will be available. SAM wouldn't touch the value of AllowOrigin you specify. So you can write anything that HTTP will accept

**Comment by hoegertn** [comment]
Is there any ETA for this feature? This will make the use of SAM so much easier...

**Comment by gertjvr** [comment]
Any updates really want this feature? 

Did a comparison between serverless framework and SAM and the only reason serverless has less lines in yaml is due to the verbose inline swagger definitions.

**Comment by ahmadsn26** [comment]
Is there any update regarding this ... i think most of us are desperately waiting for this feature.

**Comment by dinvlad** [comment]
May we also ask to be able to specify allowed methods per-resource? E.g. some resources may only have Get or only Post, while others a broader spectrum (e.g. Get, Patch, and Delete). It wouldn't make sense to return all "legal" verbs at once API-wide when only some of them are implemented, depending on resource. Unfortunately, doing that is common in other frameworks (e.g. Express), even though it goes against the RFC and the intended purpose of the Options request.

**Comment by sanathkr** [comment]
@dinvlad we can already infer the method name from the API definition and apply Cors only on it. Doable.

As for ETA, I don't have a concrete date yet. We have started the implementation. So this is definitely happening :-)

**Comment by john** [comment]
Awesome, great to hear Cors support is on the roadmap. Any further updates on ETA?

**Comment by lskrajny** [comment]
+1 Any updates maybe?

**Comment by sanathkr** [comment]
Quick update: We are almost done with the implementation. We will rollout this change shortly after QA. A couple more weeks, may be. Hang tight :-)

**Comment by mbailey** [comment]
Any update @sanathkr ?

We've implemented a swagger file to get around for the time being but are keen to remove it.

**Comment by dinvlad** [comment]
@sanathkr in the proposal, you mention a _list_ of origins. Does that mean SAM will enable dynamic generation of `Access-Control-Allow-Origin` based on the `Origin` of request? Or will it just set it to a single static value?

**Comment by bslatner** [comment]
Any updates on this? As it happens, I need it desperately :)

**Comment by prafulmehrotra** [comment]
Any update on ETA for this feature?

**Comment by lskrajny** [comment]
Are there any updates?

**Comment by sowmitranalla** [comment]
Any updates on this? It would be such a wonderful feature to have.

**Comment by frehner** [comment]
The Lambda team did an AMA on Reddit, and someone asked about updates on CORS. Here's their response:

> I hear you. There are times when we have to make tradeoffs on getting one feature out before the other. We are almost done with implementing the CORS feature now. So hang tight, it will be out real soon - Sanath

[Source](https://www.reddit.com/r/aws/comments/7a5anz/we_are_the_aws_lambda_serverless_team_ask_the/dp7b3jh/)

**Comment by sowmitranalla** [comment]
Thanks for the update @frehner ! Appreciate it!

**Comment by awHamer** [comment]
any updated on this? Thanks!

**Comment by jackmusick** [comment]
Also looking for an update... Sounds like it was going to be just a couple of weeks back in July.

**Comment by domZippilli** [comment]
+1 to this feature request. It's not SO bad to do it other ways but it's a lot of detail when this transform allows you to otherwise be so succinct.

**Comment by jarrettj** [comment]
+1 This is so damn frustrating. Been stuck on configuring CORS for an eternity! I feel lost in a rabbit hole of poo. 

With more and more applications being written in a js framework I would think that hosting a static site on S3 and speaking to a lambda function via api gateway would be well documented by now?

Will await this feature deploy and continue searching the web for a workaround.

**Comment by sdmolloy** [comment]
+1

**Comment by sanathkr** [comment]
Hey guys, apologies for the delay. We had a context switch to get [safe Lambda deployment capabilites](https://github.com/awslabs/serverless-application-model/blob/master/docs/safe_lambda_deployments.rst) in SAM. 

We know how important this feature is to all of you. It is definitely one of the top items on our roadmap. We are currently exploring options to implement in a manner that will work for both Implicit APIs & Explicit APIs (ie. where you provide your Swagger). I will keep the thread engaged as we have more developments.

Meanwhile here is the example file that tells you how to enable CORS yourself if you use Swagger - https://github.com/awslabs/serverless-application-model/tree/master/examples/2016-10-31/api_swagger_cors

**Comment by hortonsvader** [comment]
Is there any workaround in the short term?  I'm getting rejected at every OPTIONS call, and would prefer like to be unblocked even if it's a bit more work :)

**Comment by jfuss** [comment]
We created a parent task to track this. Please see #248 for updates.

Closing as duplicate of #248

**Comment by rabowskyb** [comment]
Just in case it might be of some use, here's an example of SAM with CORS enabled: 
 https://github.com/aws-samples/startup-kit-serverless-workload.

**Comment by praetp** [comment]
The example by 'ghost' above works for me but now I get 401's as I have configured a default authorizer. However, this should only be applied for the non-OPTIONS methods. For the OPTIONS, I want authorizer NONE. Anyone got a clue how to get it working ?
